쥬스 메이커 [STEP1] Kiseok, 쥬봉이

jyubong opened the pull request:
@Tediousday93 
안녕하세요. Kiseok, 쥬봉이입니다. STEP1 PR보냅니다. 잘부탁드립니다:)

---
### 코드를 작성할 때 고민되었던 점
fruitstore와 fruit 타입을 뭘로 할지 고민이 되었습니다.
- fruitstore를 한 곳으로만 할지 여러 곳으로 만들 수 있도록 할지
-> 여러 fruitstore를 만들 수 있도록 결정 따라서 struct 사용
- 1)과일을 나타낼 `fruit를 객체`로 만들어 store에 넣을지, 2)`enum 타입으로 만들어 딕셔너리`를 표현할지 고민
-> struct나 class로 객체를 만들때 재고 외에 필수적으로 필요한 프로퍼티와 메서드가 없는 것 같아 사용의 필요성이 적음
-> enum 타입이 fruit의 종류를 직관적으로 볼 수 있고, 딕셔너리의 key에는 enum을 value에는 재고를 넣어 관리를 해주면 더 나을 것 같아서 선택

### 해결이 되지 않은 점 & 조언을 얻고 싶은 부분
FruitStore의 decreaseStock메서드는 과일 재고를 확인하고 문제가 없으면 과일 수량을 낮추는 함수입니다.
과일을 2개 소비해야할때 두 과일의 재고를 먼저 확인하고 수량을 낮추어주어야하는데, 이를 구현해보니 `for-loop 2개`를 사용하게 되었습니다. for-loop나 아니면 다른 방식으로라도 `한번에 구현할 수 있는 방법`이 있을것 같은데 어렵습니다.
이부분에 대해서 조언을 듣고 싶습니다.

carti1108 committed:
feat: fruit와 juicemenu enum생성

jyubong committed:
feat: FruitStore 메서드 추가

checkStock와 decreaseStock 메서드 추가
fruits 프로퍼티와 이니셜라이저 추가

carti1108 committed:
feat: juiceMaker프로퍼티와 메소드 생성

makeJuice, checkRecipe 메소드 생성

jyubong committed:
style: created by 주석 수정

Tediousday93 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,44 @@
 // 과일 저장소 타입
-class FruitStore {
+struct FruitStore {
+    private var fruits: [Fruit: Int] = [:]
```
Dictionary의 key 자리에 설정할 타입은 `Hashable` 프로토콜을 채택해야한다는 사실 알고 계신가요? 🤔
이번에 정의하신 열거형 `Fruit`는 Hashable을 채택하지 않았음에도 컴파일 에러가 발생하지 않는 이유는 뭘까요? 💭

carti1108 replied:
enum타입에 연관값이 없다면 자동으로 hashable해집니다.
For an enum, all its associated values must conform to Hashable. (An enum without associated values has Hashable conformance even without the declaration.)
https://developer.apple.com/documentation/swift/hashable

Tediousday93 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,44 @@
+        for fruit in Fruit.allCases {
+            self.fruits[fruit] = initialStock
+        }
+    }
```
이니셜라이저를 통해 재고의 초기값을 설정해주셨군요!
CaseIterable의 활용도 좋습니다 👍👍👍

Tediousday93 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,44 @@
+        }
+    }
+    
+    private func checkStock(fruit: Fruit, quantity: Int) -> Bool {
```
과일 재고를 확인하는 메서드를 잘 만들어주셨네요 👍

Bool 타입 반환값을 갖는 메서드/프로퍼티의 이름은 질문형/확인형(assertion)으로 지어주는 편이 나중에 호출할 때 가독성이 좋다고 생각합니다.
[Swift API Design Guideline](https://www.swift.org/documentation/api-design-guidelines/#strive-for-fluent-usage)의 내용에도 아래와 같은 내용이 있으니 참고해서 이름을 바꿔볼까요?
> * Uses of Boolean methods and properties should read as assertions about the receiver when the use is nonmutating, e.g. x.isEmpty, line1.intersects(line2).

jyubong replied:
checkStrock에서 isValidStock으로, decreaseStock에서 isDecreasedStock으로 변경하였습니다.

Tediousday93 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,44 @@
+    }
+    
+    private func checkStock(fruit: Fruit, quantity: Int) -> Bool {
+        guard let stock = fruits[fruit], stock > quantity else {
```
지금은 남은 재고와 레시피의 소비량이 같은 경우, 재고가 부족하다고 나올 것 같네요 👀

jyubong replied:
stock >= quantity로 수정하였습니다.

Tediousday93 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,44 @@
+        }
+        
+        return true
+    }
```
Dictionary 전체를 돌면서 전달된 과일의 재고를 한 번에 변경해주고 있네요! 👍
fruitsInStore라는 파라미터 이름을 봤을 때, FruitStore 타입 내부의 fruits 딕셔너리를 나타내고 있는 것처럼 느껴져요.

* 호출부에서 `recipe`를 전달해주고 있으니 파라미터 이름도 전달인자에 맞춰주면 좋겠네요.
* 반복문이 2회 사용되는 부분 - 현재 레시피를 받아서 재고를 체크하는 기능과 재고를 감소시키는 기능이 같이 있기 때문에 이를 분리하고 외부에서 따로 호출해주면 해결이 될 것 같습니다. 💭
* for-in 구문을 합치게 되면 레시피 중 2번째로 검사하는 과일 재고가 부족할 경우 먼저 검사한 과일의 재고는 줄어들기 때문에 합치기는 어렵겠네요.

jyubong replied:
파라미터 이름 fruitsInStore에서 recipe로 변경하였습니다.
그리고 기존 isDecreasedStock내 재고확인 for-in구문이 isValidStock의 역할과 비슷한 것 같아 해당 로직을 isValidStock 옮겨 수정하였습니다.

Tediousday93 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -1,12 +1,48 @@
 
 // 쥬스 메이커 타입
 struct JuiceMaker {
+    private var fruitStore: FruitStore = FruitStore(initialStock: 10)
```
JuiceMaker와 FruitStore 타입을 정의할 때 struct를 선택하신 이유가 있을까요~?

jyubong replied:
FruitStore를 여러개 만들 수 있게 struct로 한 이유는 JuiceMaker하나에 FruitStore 한개가 할당되는 1:1 관계로 생각하였습니다. 그리고 JuiceMaker와 FruitStore는 class의 상속이나 참조타입을 활용하지 않고 struct로써의 단독적인 value type일 필요가 있을 것 같아 선택하였습니다.
말씀해주신 부분을 고민해보다 FruitStore가 여러개일 필요가 없을 것 같아 싱글톤으로 변경하였습니다.

그러면서 생긴 고민이
1) 지금 상황은 여러개 JuiceMaker에 1개의 FruitStore를 가지게 된 상황인데 이것 또한 맞는 것 같고
2) 이 프로젝트에서는 JuiceMaker도 1개, FruitStore도 1개만 필요할 것 같아서 JuiceMaker도 싱글톤으로 해도 되지 않을까?
라는 생각이 들었습니다. 이부분에 대해서는 계속 고민이 됩니다.

Tediousday93 replied:
JuiceMaker 어플리케이션에서 각 타입의 인스턴스가 어떻게 활용되는지 생각하면 답을 결정하실 수 있을거에요 😊

우리의 어플리케이션에서 관리하는 과일 저장소 fruitStore가 싱글톤이라면, 모든 화면에서 재고에 손쉽게 접근할 수 있다는 장점이 있습니다. (싱글톤은 대신 단점도 많습니다. 나중에 꼭 알아두셔야해요!)
만약, 화면마다 별도의 과일 저장소를 가져야 한다면 싱글톤으로 만들지 않아야겠죠?

또한 JuiceMaker의 경우 과일 저장소를 프로퍼티로 가지면서 주스를 만들고 저장소를 관리하는 기능을 제공합니다. 어떤 화면에서는 주스메이커의 메서드가 필요하지 않을 수도 있다고 생각해봅시다. 
인스턴스를 어떻게 제공하는게 좋을까요?

이번 스텝에서는 view - model이 연결되어있지 않기 때문에 파악하기 어려울 수 있습니다.
이 부분이 고민되신다면 다음 스텝에서 수정해봐도 좋을 것 같아요!

Tediousday93 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -1,12 +1,48 @@
+        }
+        
+        return recipe
+    }
```
레시피를 JuiceMenu가 알고 있으면 더 좋을 것 같네요!
enum 내부에 연산 프로퍼티를 활용해서 recipe를 정의해두면 훨씬 깔끔해질 것 같습니다 👍

carti1108 replied:
checkRecipe메소드 내부의 switch문을 JuiceMenu열거형 내에 연산 프로퍼티로 변경하였습니다.

Tediousday93 reviewed (changes requested):
기석, 쥬봉이 안녕하세요 😊
step1 고생하셨습니다.

제가 드리는 코멘트는 정답이 아닙니다. PR을 통해 저와 함께 고민하고 여러분이 스스로 코드를 작성한 이유를 설명할 수 있게 되는 것을 목표로 하면 좋겠습니다. 합당한 이유를 말씀해주시면 코멘트는 참고만 하고 넘어가셔도 좋습니다! 👍

3주간 함께 성장하는 시간이 되었으면 좋겠네요. 
잘 부탁드립니다☺️ 화이팅!! 💪💪💪

---

✅ 고민했던 점
* fruitStore를 여러 개 만들 수 있도록 결정하신 이유를 설명해주지 않으셔서 아쉽네요 🥲
* 열거형과 딕셔너리를 통해 관리해주는 방법 너무 좋습니다. 👍 열거형을 딕셔너리의 key에 활용하면 할당할 값이 Fruit 타입으로 한정되기 때문에 훨씬 관리하기 용이하다고 생각합니다.

</br>

✅ 해결이 되지 않은 점 / 조언을 얻고 싶은 부분
* 코드 코멘트에 남겨드리도록 하겠습니다.

carti1108 committed:
refactor: fruitStore 메소드 이름 수정, 재고 확인 연산자 수정

checkStrock에서 isValidStock으로 변경
decreaseStock에서 isDecreasedStock으로 변
isValidStock 비교 연산자 수정

jyubong committed:
refactor: isDecreasedStock 메서드 기능 분리, 파라미터 수정

isDecreasedStock내 재고확인 로직을 isValidStock으로 옮김
파라미터 이름 fruitsInStore에서 recipe로 변경

jyubong committed:
refactor: FruitStore를 싱글톤으로 변경

carti1108 committed:
refactor: checkRecipe메소드를 JuiceMenu내에 연산 프로퍼티로 변경

jyubong committed:
style: 컨벤션 수정, 주석 제거

Tediousday93 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -1,12 +1,44 @@
+        }
+    }
+    
+    private func isValidStock(recipe: [Fruit: Int]) -> Bool {
+        for (fruit, quantity) in recipe {
+            guard let stock = fruits[fruit], stock >= quantity else {
+                return false
+            }
+        }
+        
+        return true
+    }
+    
+    func isDecreasedStock(recipe: [Fruit: Int]) -> Bool {
+        guard isValidStock(recipe: recipe) else {
+            return false
+        }
+        
+        for (fruit, quantity) in recipe {
+            guard let stock = fruits[fruit] else {
+                return false
+            }
+            fruits[fruit] = stock - quantity
+        }
+        
+        return true
+    }
 }
```
메서드 이름을 잘 바꿔주셨네요!

여기에서 고려해볼 점은 "**기능 분리**" 입니다.
isDecreasedStock 메서드는 내부에서 stock validation을 진행한 뒤 재고를 감소시키고 재고가 감소 되었는지 되지 않았는지 확인까지 하는 3가지 기능을 한 번에 하고 있다고 생각이 됩니다.
메서드는 하나의 기능만을 하도록 하고 호출부에서 적절한 메서드를 따로 호출해주는 것이 좋다고 생각합니다. 🤔
추가로, 메서드에 **전달인자 레이블**을 활용해 더 자연스러운 영문처럼 읽힐 수 있도록 하면 좋겠네요. 

isValidStock을 internal 접근수준으로 하여 외부에서 validation을 진행하고 결과에 따라 decreaseStock을 호출해주면 좋겠네요!

carti1108 replied:
isDecreasedStock -> decreaseStock 재고만 감소하도록 수정, juiceMaker에서 isValidStock 확인 후 decreaseStock 호출하는 방식으로 기능 분리하였습니다.

Tediousday93 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -1,12 +1,23 @@
+        }
+    }
 }
+
+
+
+
```
불필요한 개행은 지워주세요!
컨벤션을 맞추는 것도 중요하고, git이 추적하는 변경사항이 발생하게 됩니다!

carti1108 replied:
수정완료했습니다!

Tediousday93 commented on JuiceMaker/JuiceMaker/Model/JuiceMenu.swift:
```diff
@@ -0,0 +1,35 @@
+    case strawberryBananaJuice
+    case mangoKiwiJuice
+    
+    var recipe: [Fruit: Int] {
+        switch self {
+        case .strawberryJuice:
+            return [.strawberry: 16]
+        case .bananaJuice:
+            return [.banana: 2]
+        case .pineappleJuice:
+            return [.pineapple: 2]
+        case .kiwiJuice:
+            return [.kiwi: 3]
+        case .mangoJuice:
+            return [.mango: 3]
+        case .strawberryBananaJuice:
+            return [.strawberry: 10, .banana: 1]
+        case .mangoKiwiJuice:
+            return [.mango: 2, .kiwi: 1]
+        }
+    }
```
연산프로퍼티로 잘 바꿔주셨군요 👍👍

Tediousday93 reviewed (changes requested):
잘 수정해주셨네요!
아래 내용만 추가로 수정해주시면 머지하도록 하겠습니다 😊
* isDecreasedStock 메서드 기능 분리
* 불필요한 개행 제거

jyubong committed:
docs: Create README.md

jyubong committed:
docs:Update README.md

jyubong committed:
docs: Update README.md

jyubong committed:
refactor: 과일 재고 감소 시키는 함수 기능 분리

isDecreasedStock -> decreaseStock 재고만 감소하도록 수정
juiceMaker에서 isValidStock 확인 후 decreaseStock 호출

jyubong committed:
style: 개행과 오타 수정

Tediousday93 commented:
고생하셨습니다 👏👏👏

다음 스텝도 화이팅!! 💪
